spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java

```diff
@@ -0,0 +1,649 @@
+    int parallelism = Math.min(partitions.size(), spark.sessionState().conf().parallelPartitionDiscoveryParallelism());
+    int numShufflePartitions = spark.sessionState().conf().numShufflePartitions();
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(targetTable.properties());
+
+    JavaSparkContext sparkContext = JavaSparkContext.fromSparkContext(spark.sparkContext());
+    JavaRDD<SparkPartition> partitionRDD = sparkContext.parallelize(partitions, parallelism);
+
+    Dataset<SparkPartition> partitionDS = spark.createDataset(
+        partitionRDD.rdd(),
+        Encoders.javaSerialization(SparkPartition.class));
+
+    List<ManifestFile> manifests = partitionDS
+        .flatMap((FlatMapFunction<SparkPartition, DataFile>) sparkPartition ->
+                listPartition(
+                    sparkPartition,
+                    spec,
+                    serializableConf,
+                    metricsConfig).iterator(),
+            Encoders.javaSerialization(DataFile.class))
+        .repartition(numShufflePartitions)
+        .map((MapFunction<DataFile, Tuple2<String, DataFile>>) file ->
+                Tuple2.apply(file.path().toString(), file),
+            Encoders.tuple(Encoders.STRING(), Encoders.javaSerialization(DataFile.class)))
+        .orderBy(col("_1"))
+        .mapPartitions(
+            (MapPartitionsFunction<Tuple2<String, DataFile>, ManifestFile>) fileTuple ->
+                buildManifest(
+                    serializableConf,
+                    spec,
+                    stagingDir,
+                    fileTuple),
+            Encoders.javaSerialization(ManifestFile.class))
+        .collectAsList();
+
+    try {
+      boolean snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+          targetTable.properties(),
+          TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+          TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+
+      AppendFiles append = targetTable.newAppend();
+      manifests.forEach(append::appendManifest);
+      append.commit();
+
+      if (!snapshotIdInheritanceEnabled) {
+        // delete original manifests as they were rewritten before the commit
+        deleteManifests(targetTable.io(), manifests);
+      }
+    } catch (Throwable e) {
+      deleteManifests(targetTable.io(), manifests);
+      throw e;
+    }
+  }
+
+  private static void deleteManifests(FileIO io, List<ManifestFile> manifests) {
+    Tasks.foreach(manifests)
+        .noRetry()
+        .suppressFailureWhenFinished()
+        .run(item -> io.deleteFile(item.path()));
+  }
+
+  /**
+   * Class representing a table partition.
+   */
+  public static class SparkPartition implements Serializable {
+    private final Map<String, String> values;
+    private final String uri;
+    private final String format;
+
+    public SparkPartition(final Map<String, String> values, final String uri, final String format) {
```

Comment:
nit: let's remove `final` from the args to be consistent with other places as we don't normally mark args as final and checkstyle should already catch if we try to modify args.